### PR TITLE
pkg/semtech-loramac: provide basic persistence for MAC state

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR provides an automatic persistence mecanism of the LoRaMAC session keys and devaddr. This allows to put a device a standby (no ram retention) when using OTAA activation and to continue sending data after wakep without having to join the network again.

Only devaddr, nwkskey, appskey and joined status are saved. But maybe we can extend to more information (DR, ADR, etc) ?

To avoid conflicts, note that this PR is based on #11541 which also provides a useful rework

### Testing procedure

- Build and flash `tests/pkg_semtech-loramac` application, join a network using OTAA procedure, send a message:
```
> loramac set deveui <deveui>
> loramac set appeui <appeui>
> loramac set appkey <appkey>
> loramac set dr 5
> loramac join otaa
Join procedure succeeded!
> loramac tx test
Message sent with success
```
- Save the current status of the MAC and reboot:
```
> loramac save
> reboot
```
- Try to send another message without joining again:
```
> loramac tx anothertest
Message sent with success
```

All steps should work. Other things that can also be tested: `uncnf`, receive a downlink. Also test that ABP activation still works.

### Issues/PRs references

Based on #11541

### CODE REVIEW

File: pkg/semtech-loramac/contrib/semtech_loramac.c [hunk 1]
```diff
@@ -406,7 +444,7 @@ void semtech_loramac_erase_config(void)
                   LORAMAC_DEVEUI_LEN + LORAMAC_APPEUI_LEN +
                   LORAMAC_APPKEY_LEN + LORAMAC_APPSKEY_LEN +
                   LORAMAC_NWKSKEY_LEN + LORAMAC_DEVADDR_LEN +
-                  uplink_counter_len);
+                  uplink_counter_len + 1);
```
Comment (fjmolinas) [inline pkg/semtech-loramac/contrib/semtech_loramac.c]: Why is this +1, is it the joined bit? Can you use a define?

**Comment by aabadie** [comment]
> when joining the join status isn't directly saved to eeprom, but only upon an eeprom_save. I think this should be done automatically just as for the uplink counter.

That's an interesting idea but it will change quite a bit the original default behaviour (where euis, keys and join state are reset after a reboot if not persisted already). I prefer having this information to be explicitly stored by the user but that's subjective.

**Comment by aabadie** [comment]
> After joining and saving the join status, I rebooted the node and tried to transmit right away. The node hanged for about 4min before getting a "Message sent with success"

At least it worked in the end :)
Did the server receive the message after 4 minutes as well ?

**Comment by fjmolinas** [comment]
> At least it worked in the end :)
Did the server receive the message after 4 minutes as well ?

Yes

**Comment by aabadie** [comment]
> Yes

Like this I would say there's something in the MAC with timers but I'm not sure.

**Comment by fjmolinas** [comment]
> That's an interesting idea but it will change quite a bit the original default behaviour (where euis, keys and join state are reset after a reboot if not persisted already). I prefer having this information to be explicitly stored by the user but that's subjective.

Ok, I won't insist on the change. Can you document it then? How the data persistence works by default, how to change it (i.e. the fact that only the uplink counter is actively saved, the fact that its only if eeprom is available, etc.).

**Comment by aabadie** [comment]
This one also needs some rebase...

**Comment by aabadie** [comment]
@fjmolinas, rebased and doc added. I couldn't reproduce your issue regarding the TX delay after reboot.

**Comment by fjmolinas** [comment]
> I couldn't reproduce your issue regarding the TX delay after reboot.

I'm still getting the error following these exact steps:

While not joined, as expected it failes to send:

```
> loramac set appeui 70B3D57ED000B02F
2019-05-29 18:12:15,452 - INFO #  loramac set appeui 70B3D57ED000B02F
> loramac set appkey 9623588BA831CD41899AE5ED63B98D37
2019-05-29 18:12:19,636 - INFO #  loramac set appkey 9623588BA831CD41899AE5ED63B98D37
> loramac set deveui 006225DDA97062F1
2019-05-29 18:12:22,907 - INFO #  loramac set deveui 006225DDA97062F1
> loramac tx hello cnf
2019-05-29 18:12:27,321 - INFO #  loramac tx hello cnf
2019-05-29 18:12:27,323 - INFO # Cannot send: not joined
```

Join, send and save:

```
> loramac join otaa
2019-05-29 18:12:35,049 - INFO #  loramac join otaa
2019-05-29 18:12:44,415 - INFO # Join procedure succeeded!
> loramac tx hello cnf
2019-05-29 18:12:52,122 - INFO #  loramac tx hello cnf
2019-05-29 18:12:55,633 - INFO # Data received: AAA, port: 1
2019-05-29 18:12:55,635 - INFO # Received ACK from network
2019-05-29 18:12:55,637 - INFO # Message sent with success
```

Reboot and try to send again, hangs for around 4'30''

```
> reboot
2019-05-29 18:13:28,560 - INFO #  reboot
2019-05-29 18:13:28,576 - INFO # main(): This is RIOT! (Version: 2019.07-devel-518-gf0165-pr-11552)
2019-05-29 18:13:28,638 - INFO # All up, running the shell now
loramac tx hello cnf
2019-05-29 18:13:30,234 - INFO #  loramac tx hello cnf
2019-05-29 18:18:00,961 - INFO # Message sent with success

```

**Comment by aabadie** [comment]
@fjmolinas, I could reproduce and there is indeed a problem that is not in master. It seems that more parameters of the MAC must stored/restored to have a fully functional device after reboot.

**Comment by aabadie** [comment]
@fjmolinas, I found the problem: the RX datarate is changed by the server after the join procedure, so one also has to persist this information to be able to receive messages from the server after a reboot.
On TTN, the DR is set to 0 that's why messages are not received. I also noticed that with high datarates, RX2 is not used, only RX1. RX2 is used with low datarates, e.g DR0, the default. That's why it was working when setting DR5.

The fix is pushed so you can try again :)

**Comment by fjmolinas** [comment]
Go!

**Pull request merged**